packages/lesswrong/server/crossposting/handlers.ts

```diff
@@ -0,0 +1,186 @@
+    if (!parsedResponse.success) {
+      // Return a 200 to avoid this triggering a health check failure,
+      // the error is still handled on the server making the request
+      return onRequestError(req, res, path, 200, "Invalid cross-site response body");
+    }
+
+    return res.status(200).send(parsedResponse.data);
+  });
+}
+
+export const addV2CrosspostHandlers = (app: Application) => {
+  addHandler(
+    app,
+    generateTokenRoute,
+    async function generateTokenCrosspostHandler({currentUser}, _payload) {
+      if (!currentUser) {
+        throw new UnauthorizedError();
+      }
+      validateCrosspostingKarmaThreshold(currentUser);
+      const token = await connectCrossposterToken.create({
+        userId: currentUser._id,
+      });
+      return {token};
+    },
+  );
+
+  addHandler(
+    app,
+    connectCrossposterRoute,
+    async function connectCrossposterCrosspostHandler(context, {
+      token,
+      localUserId,
+    }) {
+      const {userId: foreignUserId} = await connectCrossposterToken.verify(token);
+      await context.Users.rawUpdateOne({_id: foreignUserId}, {
+        $set: {fmCrosspostUserId: localUserId},
+      });
+      return {
+        status: "connected" as const,
+        foreignUserId,
+        localUserId,
+      };
+    },
+  );
+
+  addHandler(
+    app,
+    unlinkCrossposterRoute,
+    async function unlinkCrossposterCrosspostHandler(context, {token}) {
+      const {userId} = await connectCrossposterToken.verify(token);
+      await context.Users.rawUpdateOne({_id: userId}, {
+        $unset: {fmCrosspostUserId: ""},
+      });
+      return {status: "unlinked" as const};
+    },
+  );
+
+  addHandler(
+    app,
+    createCrosspostRoute,
+    async function createCrosspostHandler(context, {token}) {
+      const {
+        localUserId,
+        foreignUserId,
+        postId,
+        ...postData
+      } = await createCrosspostToken.verify(token);
+
+      const user = await context.Users.findOne({_id: foreignUserId});
+      if (!user || user.fmCrosspostUserId !== localUserId) {
+        throw new InvalidUserError();
+      }
+
+      const {data: post} = await Utils.createMutator({
+        document: {
+          userId: user._id,
+          fmCrosspost: {
+            isCrosspost: true,
+            hostedHere: false,
+            foreignPostId: postId,
+          },
+          ...postData,
+        },
+        collection: context.Posts,
+        validate: false,
+        currentUser: user,
+        context: {
+          ...context,
+          isFMCrosspostRequest: true,
+        },
+      });
+
+      return {
+        status: "posted" as const,
+        postId: post._id,
+      };
+    },
+  );
+
+  addHandler(
+    app,
+    updateCrosspostRoute,
+    async function updateCrosspostHandler(context, {token}) {
+      const {postId, ...postData} = await updateCrosspostToken.verify(token);
+      await context.Posts.rawUpdateOne({_id: postId}, {$set: postData});
```

Comment:
@oetherington Actually I feel confused about this - once we start passing in `originalContents`, doesn't that need to get saved on a Revision? So in the create crosspost handler above, it should be creating a new Revision, and in here we may have to create a new Revision if the post body changes?

Reply:
Yep, you're correct - both about what it should be doing, and that it's currently broken 😅. I just pushed some changes to part 2 that will fix it.